File: samples/python/undistort_demo.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Simple undistortion demo.

Read calibration_params.json and one image, then write undistorted output.
"""

import argparse
import json
from pathlib import Path

import cv2
import numpy as np


def _clip(v: float, lo: float, hi: float) -> float:
    return float(np.clip(float(v), lo, hi))


def undistort_with_params(
    img_bgr: np.ndarray,
    calib: dict,
    balance: float | None = None,
    out_scale: float | None = None,
    focal_scale: float | None = None,
    cx_offset: float | None = None,
    cy_offset: float | None = None,
    crop_enable: bool | None = None,
    crop_x: float | None = None,
    crop_y: float | None = None,
    crop_w: float | None = None,
    crop_h: float | None = None,
) -> np.ndarray:
    h, w = img_bgr.shape[:2]
    preview = calib.get("preview_params") or {}

    out_scale = _clip(out_scale if out_scale is not None else preview.get("out_scale", 1.0), 0.5, 3.0)
    focal_scale = _clip(focal_scale if focal_scale is not None else preview.get("focal_scale", 1.0), 0.4, 2.5)
    out_w = max(1, int(round(w * out_scale)))
    out_h = max(1, int(round(h * out_scale)))
    out_size = (out_w, out_h)

    cx_offset = _clip(cx_offset if cx_offset is not None else preview.get("cx_offset", 0.0), -out_w, out_w)
    cy_offset = _clip(cy_offset if cy_offset is not None else preview.get("cy_offset", 0.0), -out_h, out_h)

    crop_enable = bool(preview.get("crop_enable", False) if crop_enable is None else crop_enable)
    crop_x = _clip(crop_x if crop_x is not None else preview.get("crop_x", 0.05), 0.0, 0.99)
    crop_y = _clip(crop_y if crop_y is not None else preview.get("crop_y", 0.05), 0.0, 0.99)
    crop_w = _clip(crop_w if crop_w is not None else preview.get("crop_w", 0.90), 0.01, 1.0 - crop_x)
    crop_h = _clip(crop_h if crop_h is not None else preview.get("crop_h", 0.90), 0.01, 1.0 - crop_y)

    preferred_model = calib.get("preferred_model")
    if preferred_model not in ("standard", "fisheye"):
        preferred_model = "fisheye" if calib.get("fisheye_K") and calib.get("fisheye_D") else "standard"

    if preferred_model == "fisheye" and calib.get("fisheye_K") and calib.get("fisheye_D"):
        balance = _clip(
            balance if balance is not None else preview.get("balance", calib.get("fisheye_balance", 0.6)),
            0.0,
            1.0,
        )
        K = np.array(calib["fisheye_K"], dtype=np.float64)
        D = np.array(calib["fisheye_D"], dtype=np.float64).reshape(-1, 1)
        nK = cv2.fisheye.estimateNewCameraMatrixForUndistortRectify(
            K, D, (w, h), np.eye(3), None, balance, out_size
        )
        nK[0, 0] *= focal_scale
        nK[1, 1] *= focal_scale
        nK[0, 2] += cx_offset
        nK[1, 2] += cy_offset
        m1, m2 = cv2.fisheye.initUndistortRectifyMap(K, D, np.eye(3), nK, out_size, cv2.CV_16SC2)
        undist = cv2.remap(img_bgr, m1, m2, interpolation=cv2.INTER_LINEAR, borderMode=cv2.BORDER_CONSTANT)
    elif calib.get("K") and calib.get("dist"):
        K = np.array(calib["K"], dtype=np.float64)
        D = np.array(calib["dist"], dtype=np.float64)
        nK, _ = cv2.getOptimalNewCameraMatrix(K, D, (w, h), 0.0, out_size)
        nK[0, 0] *= focal_scale
        nK[1, 1] *= focal_scale
        nK[0, 2] += cx_offset
        nK[1, 2] += cy_offset
        undist = cv2.undistort(img_bgr, K, D, None, nK)
    else:
        raise ValueError("Calibration parameters are incomplete")

    if crop_enable:
        x0 = int(np.clip(round(crop_x * out_w), 0, undist.shape[1] - 1))
        y0 = int(np.clip(round(crop_y * out_h), 0, undist.shape[0] - 1))
        ww = max(1, int(round(crop_w * out_w)))
        hh = max(1, int(round(crop_h * out_h)))
        x1 = int(np.clip(x0 + ww, x0 + 1, undist.shape[1]))
        y1 = int(np.clip(y0 + hh, y0 + 1, undist.shape[0]))
        undist = undist[y0:y1, x0:x1]

    return undist


def main() -> None:
    parser = argparse.ArgumentParser(description="Undistort one image using calibration_params.json")
    parser.add_argument("--calib", required=True, help="Path to calibration_params.json")
    parser.add_argument("--image", required=True, help="Path to input image")
    parser.add_argument("--output", default="", help="Path to output image (default: <image>_undistorted.jpg)")

    parser.add_argument("--balance", type=float, default=None)
    parser.add_argument("--out-scale", type=float, default=None)
    parser.add_argument("--focal-scale", type=float, default=None)
    parser.add_argument("--cx-offset", type=float, default=None)
    parser.add_argument("--cy-offset", type=float, default=None)
    parser.add_argument("--crop-enable", action="store_true", default=None)
    parser.add_argument("--crop-x", type=float, default=None)
    parser.add_argument("--crop-y", type=float, default=None)
    parser.add_argument("--crop-w", type=float, default=None)
    parser.add_argument("--crop-h", type=float, default=None)
    args = parser.parse_args()

    calib_path = Path(args.calib)
    image_path = Path(args.image)
    if not calib_path.exists():
        raise FileNotFoundError(f"Calibration file not found: {calib_path}")
    if not image_path.exists():
        raise FileNotFoundError(f"Image file not found: {image_path}")

    with calib_path.open("r", encoding="utf-8") as f:
        calib = json.load(f)

    img = cv2.imread(str(image_path), cv2.IMREAD_COLOR)
    if img is None:
        raise RuntimeError(f"Failed to read image: {image_path}")

    undist = undistort_with_params(
        img,
        calib,
        balance=args.balance,
        out_scale=args.out_scale,
        focal_scale=args.focal_scale,
        cx_offset=args.cx_offset,
        cy_offset=args.cy_offset,
        crop_enable=args.crop_enable,
        crop_x=args.crop_x,
        crop_y=args.crop_y,
        crop_w=args.crop_w,
        crop_h=args.crop_h,
    )

    out_path = Path(args.output) if args.output else image_path.with_name(f"{image_path.stem}_undistorted.jpg")
    ok = cv2.imwrite(str(out_path), undist)
    if not ok:
        raise RuntimeError(f"Failed to write output: {out_path}")

    print(f"input : {image_path}")
    print(f"output: {out_path}")
    print(f"model : {calib.get('preferred_model')}")


if __name__ == "__main__":
    main()

File: samples/cpp/undistort_demo.cpp
#include "Undistort.h"

#include <opencv2/core/persistence.hpp>
#include <opencv2/opencv.hpp>

#include <algorithm>
#include <cmath>
#include <filesystem>
#include <iostream>
#include <string>

namespace fs = std::filesystem;

struct CalibFileData {
    UndistortInit init;
    UndistortParams preview;
};

struct CliOptions {
    std::string calib_path;
    std::string image_path;
    std::string output_path;

    bool has_balance = false;
    bool has_out_scale = false;
    bool has_focal_scale = false;
    bool has_cx = false;
    bool has_cy = false;
    bool has_crop_enable = false;
    bool has_crop_x = false;
    bool has_crop_y = false;
    bool has_crop_w = false;
    bool has_crop_h = false;

    double balance = 0.0;
    double out_scale = 1.0;
    double focal_scale = 1.0;
    double cx_offset = 0.0;
    double cy_offset = 0.0;
    bool crop_enable = false;
    double crop_x = 0.05;
    double crop_y = 0.05;
    double crop_w = 0.90;
    double crop_h = 0.90;
};

static double readDouble(const cv::FileNode& node, double fallback) {
    if (node.empty()) return fallback;
    try { return (double)node; } catch (...) { return fallback; }
}

static bool readBool(const cv::FileNode& node, bool fallback) {
    if (node.empty()) return fallback;
    try { return ((int)node) != 0; } catch (...) { return fallback; }
}

static bool readMatFromNestedSeq(const cv::FileNode& node, cv::Mat& out) {
    if (node.empty() || node.type() != cv::FileNode::SEQ) return false;
    const int h = (int)node.size();
    if (h <= 0) return false;
    const cv::FileNode first = *node.begin();
    if (first.type() != cv::FileNode::SEQ) return false;
    const int w = (int)first.size();
    if (w <= 0) return false;
    out.create(h, w, CV_64F);
    int y = 0;
    for (const auto& rowNode : node) {
        if (rowNode.type() != cv::FileNode::SEQ || (int)rowNode.size() != w) return false;
        int x = 0;
        for (const auto& v : rowNode) out.at<double>(y, x++) = (double)v;
        ++y;
    }
    return true;
}

static bool readVecAsColMat(const cv::FileNode& node, cv::Mat& out) {
    if (node.empty() || node.type() != cv::FileNode::SEQ) return false;
    const int n = (int)node.size();
    if (n <= 0) return false;
    out.create(n, 1, CV_64F);
    int i = 0;
    for (const auto& e : node) out.at<double>(i++, 0) = (double)e;
    return true;
}

static bool loadCalibration(const std::string& path, CalibFileData& data) {
    // Parse project-exported JSON calibration file.
    cv::FileStorage f(path, cv::FileStorage::READ | cv::FileStorage::FORMAT_JSON);
    if (!f.isOpened()) return false;

    if (!f["preferred_model"].empty()) data.init.preferred_model = (std::string)f["preferred_model"];
    readMatFromNestedSeq(f["K"], data.init.K);
    readVecAsColMat(f["dist"], data.init.dist);
    readMatFromNestedSeq(f["fisheye_K"], data.init.fisheye_K);
    readVecAsColMat(f["fisheye_D"], data.init.fisheye_D);

    data.preview.balance = readDouble(f["fisheye_balance"], data.preview.balance);
    cv::FileNode p = f["preview_params"];
    if (!p.empty()) {
        data.preview.balance = readDouble(p["balance"], data.preview.balance);
        data.preview.out_scale = readDouble(p["out_scale"], data.preview.out_scale);
        data.preview.focal_scale = readDouble(p["focal_scale"], data.preview.focal_scale);
        data.preview.cx_offset = readDouble(p["cx_offset"], data.preview.cx_offset);
        data.preview.cy_offset = readDouble(p["cy_offset"], data.preview.cy_offset);
        data.preview.crop_enable = readBool(p["crop_enable"], data.preview.crop_enable);
        data.preview.crop_x = readDouble(p["crop_x"], data.preview.crop_x);
        data.preview.crop_y = readDouble(p["crop_y"], data.preview.crop_y);
        data.preview.crop_w = readDouble(p["crop_w"], data.preview.crop_w);
        data.preview.crop_h = readDouble(p["crop_h"], data.preview.crop_h);
    }

    if (data.init.preferred_model != "standard" && data.init.preferred_model != "fisheye") {
        data.init.preferred_model = (!data.init.fisheye_K.empty() && !data.init.fisheye_D.empty()) ? "fisheye" : "standard";
    }
    return true;
}

static void printUsage(const char* exe) {
    std::cout
        << "Usage:\n"
        << "  " << exe << " --calib <calibration_params.json> --image <input_image> [--output <out_image>]\n"
        << "       [--balance <v>] [--out-scale <v>] [--focal-scale <v>]\n"
        << "       [--cx-offset <v>] [--cy-offset <v>] [--crop-enable <0|1>]\n"
        << "       [--crop-x <v>] [--crop-y <v>] [--crop-w <v>] [--crop-h <v>]\n";
}

static bool parseArgs(int argc, char** argv, CliOptions& opt) {
    // Parse CLI flags; unknown flags are treated as errors.
    for (int i = 1; i < argc; ++i) {
        const std::string a = argv[i];
        auto needValue = [&](const std::string& key) -> std::string {
            if (i + 1 >= argc) {
                std::cerr << "Missing value for " << key << "\n";
                std::exit(2);
            }
            return argv[++i];
        };

        if (a == "--calib") opt.calib_path = needValue(a);
        else if (a == "--image") opt.image_path = needValue(a);
        else if (a == "--output") opt.output_path = needValue(a);
        else if (a == "--balance") { opt.balance = std::stod(needValue(a)); opt.has_balance = true; }
        else if (a == "--out-scale") { opt.out_scale = std::stod(needValue(a)); opt.has_out_scale = true; }
        else if (a == "--focal-scale") { opt.focal_scale = std::stod(needValue(a)); opt.has_focal_scale = true; }
        else if (a == "--cx-offset") { opt.cx_offset = std::stod(needValue(a)); opt.has_cx = true; }
        else if (a == "--cy-offset") { opt.cy_offset = std::stod(needValue(a)); opt.has_cy = true; }
        else if (a == "--crop-enable") { opt.crop_enable = (std::stoi(needValue(a)) != 0); opt.has_crop_enable = true; }
        else if (a == "--crop-x") { opt.crop_x = std::stod(needValue(a)); opt.has_crop_x = true; }
        else if (a == "--crop-y") { opt.crop_y = std::stod(needValue(a)); opt.has_crop_y = true; }
        else if (a == "--crop-w") { opt.crop_w = std::stod(needValue(a)); opt.has_crop_w = true; }
        else if (a == "--crop-h") { opt.crop_h = std::stod(needValue(a)); opt.has_crop_h = true; }
        else if (a == "--help" || a == "-h") {
            printUsage(argv[0]);
            return false;
        } else {
            std::cerr << "Unknown arg: " << a << "\n";
            printUsage(argv[0]);
            return false;
        }
    }
    return true;
}

static UndistortParams mergeParams(const UndistortParams& base, const CliOptions& opt) {
    // Override preview defaults with explicitly provided CLI args.
    UndistortParams p = base;
    if (opt.has_balance) p.balance = opt.balance;
    if (opt.has_out_scale) p.out_scale = opt.out_scale;
    if (opt.has_focal_scale) p.focal_scale = opt.focal_scale;
    if (opt.has_cx) p.cx_offset = opt.cx_offset;
    if (opt.has_cy) p.cy_offset = opt.cy_offset;
    if (opt.has_crop_enable) p.crop_enable = opt.crop_enable;
    if (opt.has_crop_x) p.crop_x = opt.crop_x;
    if (opt.has_crop_y) p.crop_y = opt.crop_y;
    if (opt.has_crop_w) p.crop_w = opt.crop_w;
    if (opt.has_crop_h) p.crop_h = opt.crop_h;
    return p;
}

static std::string buildOutputPath(const std::string& image_path, const std::string& output_path) {
    if (!output_path.empty()) return output_path;
    fs::path in_path(image_path);
    fs::path out_name = in_path.stem().string() + "_undistorted.jpg";
    return (in_path.parent_path() / out_name).string();
}

int main(int argc, char** argv) {
    CliOptions opt;
    if (!parseArgs(argc, argv, opt)) return 2;
    if (opt.calib_path.empty() || opt.image_path.empty()) {
        printUsage(argv[0]);
        return 2;
    }

    CalibFileData calib;
    if (!loadCalibration(opt.calib_path, calib)) {
        std::cerr << "Failed to load calibration file: " << opt.calib_path << "\n";
        return 1;
    }

    cv::Mat img = cv::imread(opt.image_path, cv::IMREAD_COLOR);
    if (img.empty()) {
        std::cerr << "Failed to read image: " << opt.image_path << "\n";
        return 1;
    }

    // Build runtime params from saved preview params + CLI overrides.
    const UndistortParams runtime_params = mergeParams(calib.preview, opt);
    Undistort undistorter(calib.init, runtime_params);

    cv::Mat undist;
    std::string err;
    if (!undistorter.apply(img, undist, &err)) {
        std::cerr << err << "\n";
        return 1;
    }

    const std::string out_path = buildOutputPath(opt.image_path, opt.output_path);
    if (!cv::imwrite(out_path, undist)) {
        std::cerr << "Failed to write output image: " << out_path << "\n";
        return 1;
    }

    std::cout << "input : " << opt.image_path << "\n";
    std::cout << "output: " << out_path << "\n";
    std::cout << "model : " << calib.init.preferred_model << "\n";
    return 0;
}
